serialized_data_interface/local_sdi.py

```diff
@@ -19,42 +20,49 @@ def create_new_metadata(metadata):
     return metadata
 
 
-def change_zip_file(charm_name, metadata):
-    temp_dir = f"{charm_name}.tmp"
-    charm_zip = f"{charm_name}.charm"
+def change_zip_file(charm_name, metadata, charm_path="."):
+    try:
+        temp_dir = f"{charm_path}/{charm_name}.tmp"
+        charm_zip = f"{charm_path}/{charm_name}.charm"
 
-    os.mkdir(temp_dir)
+        os.mkdir(temp_dir)
 
-    with ZipFileWithPermissions(f"{charm_name}.charm") as old_zip:
-        old_zip.extractall(path=temp_dir)
+        with ZipFileWithPermissions(charm_zip) as old_zip:
+            old_zip.extractall(path=temp_dir)
 
-    os.remove(charm_zip)
+        os.remove(charm_zip)
 
-    with ZipFile(charm_zip, "w", ZIP_DEFLATED) as new_zip:
-        for dirpath, dirnames, filenames in os.walk(temp_dir, followlinks=True):
-            dirpath = Path(dirpath)
-            for filename in filenames:
-                filepath = dirpath / filename
-                if "metadata.yaml" in filename:
-                    new_zip.writestr(
-                        str(filepath.relative_to(temp_dir)), yaml.dump(metadata)
-                    )
-                else:
-                    new_zip.write(str(filepath), str(filepath.relative_to(temp_dir)))
+        with ZipFile(charm_zip, "w", ZIP_DEFLATED) as new_zip:
+            for dirpath, dirnames, filenames in os.walk(temp_dir, followlinks=True):
+                dirpath = Path(dirpath)
+                for filename in filenames:
+                    filepath = dirpath / filename
+                    if "metadata.yaml" in filename:
+                        new_zip.writestr(
+                            str(filepath.relative_to(temp_dir)), yaml.dump(metadata)
+                        )
+                    else:
+                        new_zip.write(
+                            str(filepath), str(filepath.relative_to(temp_dir))
+                        )
 
-    rmtree(temp_dir)
+    finally:
+        rmtree(temp_dir)
```

Comment:
It's possible that `temp_dir` might have failed to create.  Either move the `os.mkdir(temp_dir)` outside of the `try`, or make the `rmtree` more lenient:

```suggestion
        rmtree(temp_dir, ignore_errors=True)
```